Source/dataframe.lisp

```diff
@@ -0,0 +1,245 @@
+                             (lambda (col-index col-name value)
+                               (declare (ignore col-index value))
+                               (setf (elt new-cols index) col-name)
+                               (incf index))))
+      (let ((index 0))
+        (flet ((copy-row (src dst &aux (index 0))
+                 (map-data-frame-cols df src col-slice
+                                      (lambda (col-index col-name value)
+                                        (declare (ignore col-index col-name))
+                                        (setf (elt dst index) value)
+                                        (incf index)))))
+          (map-data-frame-rows df row-slice
+                               (lambda (row-index row)
+                                 (declare (ignore row-index))
+                                 (let ((new-row (make-array col-num :adjustable t)))
+                                   (copy-row row new-row)
+                                   (setf (elt new-rows index) new-row)))))))))
+
+(defgeneric map-data-frame (<data-frame> %row-slice% %col-slice% function)
+  (:method ((df <raw-data-frame>) row-slice col-slice fun)
+    (map-data-frame-rows
+     df row-slice
+     (lambda (row-index row)
+       (map-data-frame-cols
+        df row col-slice
+        (lambda (col-index col-name value)
+          (funcall fun row-index row col-index col-name value)))))))
+
+(defgeneric map-data-frame-rows (<data-frame> %row-slice% function)
+  (:method ((df <raw-data-frame>) row-slice fun)
+    (assert (typep row-slice '%row-slice%) nil '<row-invalid-slice>)
+    (let ((rows (rows df)))
+      (etypecase row-slice
+        ((eql t)
+         (loop for index from 0
+               for row across rows
+               do (funcall fun index row)))
+        ((cons (eql t) %row-index%)
+         (loop for index from 0 upto (cdr row-slice)
+               for row across rows
+               do (funcall fun index row)))
+        ((cons %row-index% (eql t))
+         (loop for index from (car row-slice)
+               for row = (elt rows index)
+               do (funcall fun index row)))
+        ((cons %row-index% %row-index%)
+         (loop for index from (car row-slice) upto (cdr row-slice)
+               for row = (elt rows index)
+               do (funcall fun index row)))
+        ((vector %row-index%)
+         (loop for index across row-slice
+               for row = (elt rows index)
+               do (funcall fun index row)))))))
+
+(defgeneric map-data-frame-cols (<data-frame> %row-index% %col-slice% function)
+  (:method ((df <raw-data-frame>) row col-slice fun)
+    (assert (typep row '%row-spec%)        nil '<row-invalid-index>)
+    (assert (typep col-slice '%col-slice%) nil '<col-invalid-slice>)
+    (typecase row
+      (integer (setf row (elt (rows df) row)))
+      ;; Undocumented hack to map over col names.
+      (null (setf row (make-array (length (rows df))))))
```

Comment:
Reading an uninitialized element of an array is undefined. Maybe just change to
`
(null (setf row (make-array (length (rows df)) :initial-element nil))))
`

Reply:
fixed

Reply:
after a thought I've changed it to simply use cols (the arrity always matches and it makes sense to pass the column name as a value).